zrml/liquidity-mining/src/lib.rs

```diff
@@ -0,0 +1,255 @@
+            Ok(())
+        }
+    }
+
+    #[pallet::config]
+    pub trait Config: frame_system::Config {
+        type Currency: ReservableCurrency<Self::AccountId>;
+        type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+        type MarketId: MarketId;
+        type PalletId: Get<PalletId>;
+        type WeightInfo: WeightInfoZeitgeist;
+    }
+
+    #[pallet::event]
+    pub enum Event<T>
+    where
+        T: Config, {}
+
+    #[pallet::error]
+    pub enum Error<T> {}
+
+    #[cfg(feature = "std")]
+    #[pallet::genesis_build]
+    impl<T: Config> GenesisBuild<T> for GenesisConfig<T> {
+        fn build(&self) {
+            T::Currency::deposit_creating(&Pallet::<T>::pallet_account_id(), self.initial_balance);
+            <PerBlockIncentive<T>>::put(self.per_block_distribution);
+        }
+    }
+
+    #[cfg(feature = "std")]
+    #[derive(Debug)]
+    #[pallet::genesis_config]
+    pub struct GenesisConfig<T: Config> {
+        pub initial_balance: BalanceOf<T>,
+        pub per_block_distribution: BalanceOf<T>,
+    }
+
+    #[cfg(feature = "std")]
+    impl<T> Default for GenesisConfig<T>
+    where
+        T: Config,
+    {
+        #[inline]
+        fn default() -> Self {
+            Self {
+                initial_balance: BalanceOf::<T>::zero(),
+                per_block_distribution: BalanceOf::<T>::zero(),
+            }
+        }
+    }
+
+    #[pallet::hooks]
+    impl<T: Config> Hooks<T::BlockNumber> for Pallet<T> {
+        /// Manages incentives on each block finalization.
+        fn on_finalize(curr_block: T::BlockNumber) {
+            let _ = DistributeIncentivesBasedOnBoughtShares::<T>::exec(curr_block);
+            let _ = RemoveIncentivesBasedOnSoldShares::<T>::exec();
+        }
+    }
+
+    #[pallet::pallet]
+    pub struct Pallet<T>(PhantomData<T>);
+
+    impl<T> Pallet<T>
+    where
+        T: Config,
+    {
+        // pot/fund account
+        pub(crate) fn pallet_account_id() -> T::AccountId {
+            T::PalletId::get().into_account()
+        }
+    }
+
+    impl<T> LiquidityMiningPalletApi for Pallet<T>
+    where
+        T: Config,
+    {
+        type AccountId = T::AccountId;
+        type Balance = BalanceOf<T>;
+        type BlockNumber = T::BlockNumber;
+        type MarketId = T::MarketId;
+
+        fn add_market_period(market_id: Self::MarketId, period: [Self::BlockNumber; 2]) {
+            <Markets<T>>::insert(market_id, period)
+        }
+
+        fn add_shares(
+            account_id: Self::AccountId,
+            market_id: Self::MarketId,
+            shares: Self::Balance,
+        ) {
+            <BlockBoughtShares<T>>::mutate(market_id, account_id, |total_shares| {
+                *total_shares = total_shares.saturating_add(shares);
+            })
+        }
+
+        #[frame_support::transactional]
+        fn distribute_market_incentives(market_id: &Self::MarketId) -> DispatchResult {
```

Comment:
This needs to check that the pallet account has funds in it to distribute rewards. The pool could be zero (which it will likely start out with when we first deploy this to the testnet.

Reply:
Added `ensure_can_withdrawn` but `#[frame_support::transactional]` should theoretically take care of this scenario

Reply:
Ok, can we return early or some other way handle the case that there are no funds to make the payout of the rewards? Since when we first deploy this there will be no funds being used to pay out rewards, now this will prevent a pool from being move to stale

Reply:
`ensure_can_withdraw` will return an early error if the fund does not have enough funds. Should the `ensure_can_withdraw` check return an early `Ok` instead?
It is also possible to ignore `distribute_market_incentives` within `set_pool_to_stale` using `let _ = T::LiquidityMining::distribute_market_incentives(market_id);` instead of `T::LiquidityMining::distribute_market_incentives(market_id)?;`

Reply:
Yes, it should return an early `Ok` and also log an event `LiquidityRewardsNotDistributed` so we know when it happens. The behaviour should be either everyone gets distributed rewards or no one is distributed rewards, but either way the pool still moves to `Stale`.

Reply:
Done

As stated by https://github.com/zeitgeistpm/zeitgeist/pull/122#discussion_r636214190, I would like to create (if possible) events to this pallet in a following PR because the changes are already too big.

Reply:
Which issue did you create?

Reply:
#197